### PR TITLE
Reduce memory consumption for invocations of cached actions

### PR DESCRIPTION
**Pull request by dubeejw**
Creates new datatype called WhiskActionMetadata that drops code property from its Exec data type for action invocations. This prevents action code from being loaded into memory.

### CODE REVIEW

**Comment by dubeejw** [comment]
Thinking that all of the duplicate data types can be used as bases to extend the corresponding types that require the code property.

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala [hunk 1]
```diff
@@ -150,6 +157,18 @@ class CouchDbRestStore[DocumentAbstraction <: DocumentSerializer](
           ErrorLevel))
   }
 
+  def newRead[A](ma: Manifest[A], value: JsValue) = {
```
Comment (dubeejw) [inline common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: TODO: move this.

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 2]
```diff
@@ -126,6 +178,31 @@ protected[core] case class CodeExecAsAttachment(manifest: RuntimeManifest,
...
+    extends CodeExec2[Attachment[String]] {
+  override val kind = manifest.kind
+  override val image = manifest.image
+  override val sentinelledLogs = manifest.sentinelledLogs.getOrElse(true)
+  override val deprecated = manifest.deprecated.getOrElse(false)
+  override val pull = false
+  override lazy val binary = true
+  override def codeAsJson = "".toJson //code.toJson
+
+  /*
```
Comment (dubeejw) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: TODO: delete commetted out code if not needed.

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 3]
```diff
@@ -144,12 +221,33 @@ protected[core] case class BlackBoxExec(override val image: ImageName,
...
 }
 
+protected[core] case class BlackBoxExec2(override val image: ImageName,
+                                        override val entryPoint: Option[String],
+                                        val native: Boolean)
+    extends CodeExec2[Option[String]] {
+  override val kind = Exec2.BLACKBOX
+  override val deprecated = false
+  override def codeAsJson = "".toJson //code.toJson
+  override lazy val binary = false //TODO code map { Exec.isBinaryCode(_) } getOrElse false
```
Comment (dubeejw) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: Can't determine binary type for actions with code property removed.

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 4]
```diff
@@ -278,3 +376,132 @@ protected[core] object Exec extends ArgNormalizer[Exec] with DefaultJsonProtocol
...
+        val main = a.entryPoint.map("main" -> JsString(_))
+        JsObject(base ++ main)
+
+      case s @ SequenceExec2(comp) =>
+        JsObject("kind" -> JsString(s.kind), "components" -> comp.map(_.qualifiedNameWithLeadingSlash).toJson)
+
+      case b: BlackBoxExec2 =>
+        val base =
+          Map("kind" -> JsString(b.kind), "image" -> JsString(b.image.publicImageName), "binary" -> JsBoolean(b.binary))
+        //val code = b.code.filter(_.trim.nonEmpty).map("code" -> JsString(_))
```
Comment (dubeejw) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: TODO: delete.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 5]
```diff
@@ -222,29 +222,49 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
    */
   override def activate(user: Identity, entityName: FullyQualifiedEntityName, env: Option[Parameters])(
     implicit transid: TransactionId) = {
+
+    /*parameter('blocking ? false, 'result ? false, 'timeout.as[FiniteDuration] ? WhiskActionsApi.maxWaitForBlockingActivation) { (blocking, result, waitOverride) =>
```
Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: TODO: delete.

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 6]
```diff
@@ -222,29 +222,49 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
...
+        logging.info(this, "here1")
+        getEntity(WhiskActionMini, entityStore, entityName.toDocId, Some {
+          act: WhiskActionMini =>
+            logging.info(this, "here2")
             // resolve the action --- special case for sequences that may contain components with '_' as default package
             val action = act.resolve(user.namespace)
-            onComplete(entitleReferencedEntities(user, Privilege.ACTIVATE, Some(action.exec))) {
+            onComplete(entitleReferencedEntities2(user, Privilege.ACTIVATE, Some(action.exec))) {
               case Success(_) =>
+                logging.info(this, "here3")
```
Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: TODO: remove extra logging.

**Comment by dubeejw** [comment]
TODO: need test in ActionsApiTests that ensures code property does not exist.

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala (hunk 1 above)
Comment (dubeejw) [reply common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Test failures are a result of those hack. Need to figure out a better way.

**Comment by dubeejw** [comment]
PG2 2104 ⏳

**Comment by dubeejw** [comment]
Profiling of action invocations using the changes provided in this PR. The action is a 34 MB zipped Node action. 

screen shot 2017-09-26 at 2 46 09 pm

First action invocation is a cache miss causing the controller to fetch the action from the DB. This means the code associated with the action is download into the controller causing a spike in the used heap. Garbage collection is then performed to lower the used heap. The same action is then invoked three times without the previous heap increase since the action without the code property is cached, which does not cause a fetch from the DB.

**Comment by markusthoemmes** [comment]
@dubeejw nice view, can you provide a profiling result from before so we can compare?

**Comment by dubeejw** [comment]
PG3 1136 🔵

**Comment by dubeejw** [comment]
@rabbah, want to review?

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 7]
```diff
@@ -45,7 +45,7 @@ import whisk.core.entity.size.SizeString
  *   main  : name of the entry point function, when using a non-default value (for Java, the name of the main class)" }
  */
 sealed abstract class Exec extends ByteSizeable {
-  override def toString = Exec.serdes.write(this).compactPrint
+  override def toString: String = Exec.serdes.write(this).compactPrint
```
[the remark this answers is not in the document]
Comment (dubeejw) [reply common/scala/src/main/scala/whisk/core/entity/Exec.scala]: Would have to investigate to see when this method is used.

**Pull request merged**